src/UI/Layout.hs

```diff
@@ -25,66 +29,80 @@ offset by = wrap . Offset by
 resizeable :: (Size (Maybe a) -> Layout a b) -> Layout a b
 resizeable = wrap . Resizeable
 
+getMaxSize :: Layout a (Size (Maybe a))
+getMaxSize = wrap (Resizeable pure)
+
+measure :: Layout a b -> (Size a -> Layout a b) -> Layout a b
+measure child = wrap . Measure child
+
 newtype Stack a b = Stack { unStack :: Layout a b }
 
 stack :: (Real a, Foldable t) => t (Layout a (Size a)) -> Layout a (Size a)
 stack = unStack . foldMap Stack
 
 measureLayout :: Real a => Layout a (Size a) -> Size a
-measureLayout = fromMaybe (Size 0 0) . fitLayoutTo (pure Nothing)
+measureLayout = fromMaybe (Size 0 0) . fitLayout (pure Nothing)
+
+fitLayout :: Real a => Size (Maybe a) -> Layout a (Size a) -> Maybe (Size a)
+fitLayout = fitLayoutWith layoutSizeAlgebra
 
-fitLayoutTo :: Real a => Size (Maybe a) -> Layout a (Size a) -> Maybe (Size a)
-fitLayoutTo maxSize layout = case fromF layout of
+fitLayoutAndAnnotate :: Real a => Size (Maybe a) -> Layout a (Size a) -> ALayout a (Size a) (Maybe (Size a))
+fitLayoutAndAnnotate = fitLayoutWith (annotatingBidi layoutSizeAlgebra)
+
+layoutSizeAlgebra :: Real a => CofreerF (FreerF (LayoutF a) (Size a)) (Size (Maybe a)) (Maybe (Size a)) -> Maybe (Size a)
+layoutSizeAlgebra (Cofree maxSize runC layout) = case layout of
   Pure size | maxSize `encloses` size -> Just (fromMaybe <$> size <*> maxSize)
-  Free (Inset inset rest) | maxSize `encloses` (2 * inset) -> (2 * inset +) <$> fitLayoutTo (subtractSize (2 * inset)) (toF rest)
-  Free (Offset offset rest) | maxSize `encloses` pointSize offset -> (pointSize offset +) <$> fitLayoutTo (subtractSize (pointSize offset)) (toF rest)
-  Free (Resizeable resize) -> fitLayoutTo maxSize (toF (resize maxSize))
+  Free runF l -> case l of
+    Inset by child -> (2 * by +) <$> runC (runF child)
+    Offset by child -> (pointSize by +) <$> runC (runF child)
+    Resizeable resize -> runC (runF (resize maxSize))
+    Measure child withMeasurement -> runC (runF child) >>= runC . runF . withMeasurement
```

Comment:
A single algebra is used both to compute the size for each node and to annotate each node with its computed size.